### PR TITLE
libxml++ 2.38.0

### PR DESCRIPTION
**Pull request by tschoonj**
version bump
audit --strict compliance

### CODE REVIEW

File: Library/Formula/libxml++.rb [hunk 1]
```diff
@@ -18,6 +16,21 @@ class Libxmlxx < Formula
...
+
+      int main(int argc, char *argv[])
+      {
+         xmlpp::Document document;
+         document.set_internal_subset("homebrew", "", "http://www.brew.sh/xml/test.dtd");
+         xmlpp::Element *rootnode = document.create_root_node("homebrew");
+         return 0;
+      }
+    EOS
+    system ENV.cxx, "-I#{HOMEBREW_PREFIX}/include/libxml++-2.6", "-I#{HOMEBREW_PREFIX}/lib/libxml++-2.6/include", "-I#{HOMEBREW_PREFIX}/include/glibmm-2.4", "-I#{HOMEBREW_PREFIX}/lib/glibmm-2.4/include", "-I#{HOMEBREW_PREFIX}/include/glib-2.0", "-I#{HOMEBREW_PREFIX}/lib/glib-2.0/include", "-I#{HOMEBREW_PREFIX}/opt/gettext/include", "-I#{HOMEBREW_PREFIX}/include/sigc++-2.0", "-I#{HOMEBREW_PREFIX}/lib/sigc++-2.0/include", "-I/usr/include/libxml2", "test.cpp", "-L#{HOMEBREW_PREFIX}/lib", "-L#{HOMEBREW_PREFIX}/lib", "-L#{HOMEBREW_PREFIX}/lib", "-L#{HOMEBREW_PREFIX}/opt/gettext/lib", "-L#{HOMEBREW_PREFIX}/lib", "-lxml++-2.6", "-lxml2", "-lglibmm-2.4", "-lgobject-2.0", "-lglib-2.0", "-lintl", "-lsigc-2.0", "-o", "test"
```
Comment (xu-cheng) [inline Library/Formula/libxml++.rb]: You need to sort out the arguments. Please don't use `HOMEBREW_PREFIX`. Instead you should use `include`, `lib` and `Formula["foo"].opt_include`, `Formula["foo"].opt_lib`. And please remove all the unnecessary arguments such as `"-I#{HOMEBREW_PREFIX}/opt/gettext/include"`. Thanks.

**Comment by tschoonj** [comment]
I vote to keep them as I put them here: your suggestion won't work anyway I think:

`Formula["libxml++"].opt_include` matches `#{HOMEBREW_PREFIX}/opt/libxml++/include` right? The include statement expects the headers to be one level deeper at `libxml++-2.6`.

I generate these annoyingly long compiler expressions using the output of `pkg-config --cflags` and `--libs`, which I then transform with VIM regular expressions into what you see. This way I am sure to get all possible dependencies right (sometimes too many but that's harmless). Your approach would require a lot more manual intervention (typing) getting the compilation command right and as I said, it's not exactly fun the way it is now...

**Comment by xu-cheng** [comment]
Please don't use such expressions. `HOMEBREW_PREFIX` shouldn't be allowed in these arguments as the dependencies may not even be linked to that place. The hard coded version part would make it easy to break for future updates. At the same time, there are lots of duplicated in your expressions.  

If you find manually intervention is too hard, please use `pkg-config --cflags` directly. Thanks.

**Comment by tschoonj** [comment]
It may be possible that the test may break in the future because of using `HOMEBREW_PREFIX` but it's quite unlikely. If it would happen, I will gladly fix it. For now, this is for me the fastest way to implement fully functional tests and I would like to keep it like this, even if it's not the best possible solution. I am happy contributing to Homebrew but I want to keep it fun for me too...

About using `pkg-config`: ideally I would use that but @mikemcquaid is very much against that since he doesn't like build-time dependencies in the test-block (see the discussion in #38741). It is still my opinion that `pkg-config` should be used for these tests.

**Comment by xu-cheng** [comment]
Can you show me which `pkg-config` command in specific you are using to generate this? And what's the original output? Thanks.

**Comment by tschoonj** [comment]
`pkg-config --cflags libxml++-2.6` and pkg-config --libs libxml++-2.6

**Comment by xu-cheng** [comment]
How about the original output? (I can run it myself as I don't have gtk/x11 env installed)

**Comment by tschoonj** [comment]
````
-I/usr/local/Cellar/libxml++/2.38.0/include/libxml++-2.6 -I/usr/local/Cellar/libxml++/2.38.0/lib/libxml++-2.6/include -I/usr/local/Cellar/glibmm/2.44.0/include/glibmm-2.4 -I/usr/local/Cellar/glibmm/2.44.0/lib/glibmm-2.4/include -I/usr/local/Cellar/glib/2.44.0/include/glib-2.0 -I/usr/local/Cellar/glib/2.44.0/lib/glib-2.0/include -I/usr/local/opt/gettext/include -I/usr/local/Cellar/libsigc++/2.4.1/include/sigc++-2.0 -I/usr/local/Cellar/libsigc++/2.4.1/lib/sigc++-2.0/include -I/usr/include/libxml2
````

and 

````
-L/usr/local/Cellar/libxml++/2.38.0/lib -L/usr/local/Cellar/glibmm/2.44.0/lib -L/usr/local/Cellar/glib/2.44.0/lib -L/usr/local/opt/gettext/lib -L/usr/local/Cellar/libsigc++/2.4.1/lib -lxml++-2.6 -lxml2 -lglibmm-2.4 -lgobject-2.0 -lglib-2.0 -lintl -lsigc-2.0
````

**Comment by xu-cheng** [comment]
I tried to write a small tool you can used in here: https://gist.github.com/xu-cheng/04a353ad7b5667e4f4fe

It should help you to generate the code you can used in the test.

**Comment by xu-cheng** [comment]
@tschoonj In case you missed it (it seems that I cannot ping people in gist comment), the tool has been updated to let you manually pass the pkg-config name.

**Comment by tschoonj** [comment]
formula works after I substitute the dashes in the pkg-config-name variable name with an underscore: otherwise I get a runtime error. This is a great tool! Consider adding it to the homebrew codebase as it will be useful for many people!

Will update the formulas later tonight or tomorrow.

Thanks again,

Tom

Dr. Tom Schoonjans
Research Fellow

Bundesanstalt für Materialforschung und -prüfung (Federal Institute for Materials Research and Testing)
Richard-Willstätter-Strasse 11, 12489 Berlin
Germany

Telephone +49 (0) 30-8104-5612

Tom.Schoonjans@bam.de<mailto:Tom.Schoonjans@bam.de>
www.bam.de


On 10 May 2015, at 18:19, Xu Cheng <notifications@github.com<mailto:notifications@github.com>> wrote:


@tschoonj<https://github.com/tschoonj> In case you missed it (it seems that I cannot ping people in gist comment), the tool has been updated to let you manually pass the pkg-config name.

**Comment by tschoonj** [comment]
More comments:

1) Can you get indent the array elements and the closing bracket of flags? I get audit --strict warning about this
2) In the case of libxml++, libsigc++ is a dependency: this name is used as variable name which results in a run-time error because of the '++'. Could you perhaps transform them into 'xx'?

Thanks

**Comment by DomT4** [comment]
What's stopping `libxml` finding those paths? None of them are `keg_only` right? So all should be available along the `$PATH` natively?

**Comment by tschoonj** [comment]
Any idea what went wrong with the mountain lion build? I guess I am using some feature thats not supported by ML's old ruby?

**Comment by xu-cheng** [comment]
1. Indent part should be fixed by your editor. 
2. I have updated to substitute these characters.
3. You need to put `*flags` in the end.

**Comment by tschoonj** [comment]
@xu-cheng thanks for the fixes. You did forget to change the variable name `pkg-config-name` to `pkg_config_name`...

I fixed the formula. If it builds without errors you can pull it in.

**Comment by xu-cheng** [comment]
Thanks for pointing that out. I have fixed it.

**Comment by tschoonj** [comment]
Something got broken in your last commit.

I get this for libxml++ now:

````
ENV.libxml2
gettext = Formula["gettext"]
glib = Formula["glib"]
glibmm = Formula["glibmm"]
libsigcxx = Formula["libsigc++"]
flags = ENV["CPPFLAGS"].split + ENV["LDFLAGS"].split
flags += %W[
  -I#{gettext.opt_}
  -I#{glib.opt_}
  -I#{glibmm.opt_}
  -I#{include}/libxml++-2.6
  -I#{libsigcxx.opt_}
  -I#{lib}/libxml++-2.6/include
  -L#{gettext.opt_}
  -L#{glib.opt_}
  -L#{glibmm.opt_}
  -L#{libsigcxx.opt_}
  -L#{lib}
  -lglib-2.0
  -lglibmm-2.4
  -lgobject-2.0
  -lintl
  -lsigc-2.0
  -lxml++-2.6
  -lxml2
]
````

It chops of after the underscore in opt_include and opt_lib

**Comment by tschoonj** [comment]
actually it's worse: it chops of more than I initially thought plus it removes entire lines which should not have been

I expected this

````
flags += %W[
      -I#{gettext.opt_include}
      -I#{glib.opt_include}/glib-2.0
      -I#{glib.opt_lib}/glib-2.0/include
      -I#{glibmm.opt_include}/glibmm-2.4
      -I#{glibmm.opt_lib}/glibmm-2.4/include
      -I#{include}/libxml++-2.6
      -I#{libsigcxx.opt_include}/sigc++-2.0
      -I#{libsigcxx.opt_lib}/sigc++-2.0/include
      -I#{lib}/libxml++-2.6/include
      -L#{gettext.opt_lib}
      -L#{glib.opt_lib}
      -L#{glibmm.opt_lib}
      -L#{libsigcxx.opt_lib}
      -L#{lib}
      -lglib-2.0
      -lglibmm-2.4
      -lgobject-2.0
      -lintl
      -lsigc-2.0
      -lxml++-2.6         
      -lxml2              
    ]
````

**Comment by xu-cheng** [comment]
Sorry, I didn't test it locally. It should be fixed now.

**Comment by tschoonj** [comment]
Thanks, works like a charm now...

**Comment by tschoonj** [comment]
@xu-cheng One more problem with your script:

You cannot put `ENV.x11` in the test block: it works but it doesn't pass `audit --strict` which will complain and ask you to put instead: `depends_on :x11`

I had this problem before (#35725) and managed to circumvent it using:

````
ENV.prepend "PKG_CONFIG_PATH", OS::Mac::X11.lib/"pkgconfig"
````

You may want to update your script with this change.

Thanks

**Pull request closed without merging**